File: src/products/service.ts
import type { PrismaClient } from "@prisma/client";
import { AppError } from "../lib/middleware.js";
import type { CreateCategoryInput, CreateProductInput } from "./schema.js";

export async function createCategory(
  prisma: PrismaClient,
  data: CreateCategoryInput,
) {
  const category = await prisma.category.create({
    data: {
      name: data.name,
      description: data.description ?? null,
    },
  });
  return category;
}

export async function listCategories(prisma: PrismaClient) {
  return prisma.category.findMany();
}

export async function createProduct(
  prisma: PrismaClient,
  data: CreateProductInput,
) {
  // Verify category exists if provided
  if (data.category_id) {
    const category = await prisma.category.findUnique({
      where: { id: data.category_id },
    });
    if (!category) {
      throw new AppError(404, "Category not found");
    }
  }

  // Create product and initialize inventory in a transaction
  const product = await prisma.$transaction(async (tx) => {
    const newProduct = await tx.product.create({
      data: {
        name: data.name,
        description: data.description ?? null,
        price: data.price,
        categoryId: data.category_id ?? null,
      },
    });

    // Initialize inventory for new product (matches products/api.py:49-53)
    await tx.inventory.create({
      data: {
        productId: newProduct.id,
        quantity: 0,
        reserved: 0,
      },
    });

    return newProduct;
  });

  return product;
}

export async function getProduct(prisma: PrismaClient, productId: number) {
  const product = await prisma.product.findUnique({
    where: { id: productId },
  });
  if (!product) {
    throw new AppError(404, "Product not found");
  }
  return product;
}

export async function listProducts(prisma: PrismaClient) {
  return prisma.product.findMany();
}

File: src/ecommerce/products/service.py
"""Product and category business logic."""

from fastapi import HTTPException
from sqlmodel import select
from sqlmodel.ext.asyncio.session import AsyncSession

from ecommerce.products.models import Category, Product
from ecommerce.products.schemas import CategoryCreate, ProductCreate
from ecommerce.inventory.models import Inventory


async def create_category(session: AsyncSession, data: CategoryCreate) -> Category:
    """Create a new category."""
    category = Category(name=data.name, description=data.description)
    session.add(category)
    await session.commit()
    await session.refresh(category)
    return category


async def list_categories(session: AsyncSession) -> list[Category]:
    """List all categories."""
    result = await session.execute(select(Category))
    return list(result.scalars().all())


async def create_product(session: AsyncSession, data: ProductCreate) -> Product:
    """Create a new product with category validation and inventory initialization.

    Uses flush-then-commit for atomic product + inventory creation.
    """
    # Verify category exists if provided
    if data.category_id:
        category = await session.get(Category, data.category_id)
        if not category:
            raise HTTPException(status_code=404, detail="Category not found")

    product = Product(
        name=data.name,
        description=data.description,
        price=data.price,
        category_id=data.category_id,
    )
    session.add(product)
    await session.flush()  # Get product.id without committing

    # Initialize inventory for new product (atomic with product creation)
    inventory = Inventory(product_id=product.id, quantity=0, reserved=0)
    session.add(inventory)

    await session.commit()
    await session.refresh(product)
    return product


async def get_product(session: AsyncSession, product_id: int) -> Product:
    """Get product by ID."""
    product = await session.get(Product, product_id)
    if not product:
        raise HTTPException(status_code=404, detail="Product not found")
    return product


async def list_products(session: AsyncSession) -> list[Product]:
    """List all products."""
    result = await session.execute(select(Product))
    return list(result.scalars().all())
